tools/i2cget.8
.TH I2CGET 8 "July 2021"
.SH "NAME"
i2cget \- read from I2C/SMBus chip registers

.SH SYNOPSIS
.B i2cget
.RB [ -f ]
.RB [ -y ]
.RB [ -a ]
.I i2cbus
.I chip-address
.RI [ "data-address " [ "mode " [ length ]]]
.br
.B i2cget
.B -V

.SH DESCRIPTION
i2cget is a small helper program to read registers visible through the I2C
bus (or SMBus).

.SH OPTIONS
.TP
.B -V
Display the version and exit.
.TP
.B -f
Force access to the device even if it is already busy. By default, i2cget
will refuse to access a device which is already under the control of a
kernel driver. Using this flag is dangerous, it can seriously confuse the
kernel driver in question. It can also cause i2cget to return an invalid
value. So use at your own risk and only if you know what you're doing.
.TP
.B -y
Disable interactive mode. By default, i2cget will wait for a confirmation
from the user before messing with the I2C bus. When this flag is used, it
will perform the operation directly. This is mainly meant to be used in
scripts. Use with caution.
.TP
.B -a
Allow using addresses between 0x00 - 0x07 and 0x78 - 0x7f. Not recommended.
.PP
There are two required options to i2cget. \fIi2cbus\fR indicates the number
or name of the I2C bus to be scanned.  This number should correspond to one of
the busses listed by \fIi2cdetect -l\fR. \fIchip-address\fR specifies the
address of the chip on that bus, and is an integer between 0x08 and 0x77.
.PP
\fIdata-address\fR specifies the address on that chip to read from, and is
an integer between 0x00 and 0xFF. If omitted, the currently active register
will be read (if that makes sense for the considered chip).
.PP
The \fImode\fR parameter, if specified, is one of the letters \fBb\fP,
\fBw\fP, \fBc\fP, \fBs\fP, or \fBi\fP, corresponding to a read byte data, a read
word data, a write byte/read byte, an SMBus block read, or an I2C block read
transaction, respectively. A \fBp\fP can also be appended to the \fImode\fR
parameter to enable PEC, except for I2C block transactions. If the \fImode\fR
parameter is omitted,
i2cget defaults to a read byte data transaction, unless \fIdata-address\fR is
also omitted, in which case the default (and only valid) transaction is a
single read byte.
.PP
The \fIlength\fR parameter, if applicable and specified, sets the length
of the block transaction. Valid values are between 1 and 32. Default value
is 32.

.SH WARNING
i2cget can be extremely dangerous if used improperly. I2C and SMBus are designed
in such a way that an SMBus read transaction can be seen as a write transaction by
certain chips. This is particularly true if setting \fImode\fR to \fBcp\fP (write byte/read
byte with PEC). Be extremely careful using this program.

.SH EXAMPLES
.PP
Get the value of 8-bit register 0x11 of the I2C device at 7-bit address 0x2d
on bus 1 (i2c-1), after user confirmation:
.nf
.RS
# i2cget 1 0x2d 0x11
.RE
.fi
.PP
Get the value of 16-bit register 0x00 of the I2C device at 7-bit address 0x48
on bus 1 (i2c-1), after user confirmation:
.nf
.RS
# i2cget 1 0x48 0x00 w
.RE
.fi
.PP
Set the internal pointer register of a 24C02 EEPROM at 7-bit address 0x50
on bus 9 (i2c-9) to 0x00, then read the first 2 bytes from that EEPROM:
.nf
.RS
# i2cset -y 9 0x50 0x00 ; i2cget -y 9 0x50 ; i2cget -y 9 0x50
.RE
.fi
This assumes that the device automatically increments its internal pointer
register on every read, and supports read byte transactions (read without
specifying the register address, "Receive Byte" in SMBus terminology.)
Most EEPROM devices behave that way. Note that this is only safe as long as
nobody else is accessing the I2C device at the same time. A safer approach
would be to use a "Read Word" SMBus transaction instead, or an I2C Block
Read transaction to read more than 2 bytes.
.PP
Set the internal pointer register of a 24C32 EEPROM at 7-bit address 0x53
on bus 9 (i2c-9) to 0x0000, then read the first 2 bytes from that EEPROM:
.nf
.RS
# i2cset -y 9 0x53 0x00 0x00 ; i2cget -y 9 0x53 ; i2cget -y 9 0x53
.RE
.fi
This again assumes that the device automatically increments its internal
pointer register on every read, and supports read byte transactions. While
the previous example was for a small EEPROM using 8-bit internal addressing,
this example is for a larger EEPROM using 16-bit internal addressing. Beware
that running this command on a small EEPROM using 8-bit internal addressing
would actually \fIwrite\fR 0x00 to the first byte of that EEPROM. The safety
concerns raised above still stand, however in this case there is no SMBus
equivalent, so this is the only way to read data from a large EEPROM if your
master isn't fully I2C capable. With a fully I2C capable master, you would
use \fIi2ctransfer\fR to achieve the same in a safe and faster way.
.PP
Read the first 8 bytes of an EEPROM device at 7-bit address 0x50
on bus 4 (i2c-4):
.nf
.RS
# i2cget -y 4 0x50 0x00 i 8
.RE
.fi

.SH BUGS
To report bugs or send fixes, please write to the Linux I2C mailing list
<linux-i2c@vger.kernel.org> with Cc to the current maintainer:
Jean Delvare <jdelvare@suse.de>.

.SH SEE ALSO
i2cdetect(8), i2cdump(8), i2cset(8), i2ctransfer(8)

.SH AUTHOR
Jean Delvare

This manual page was strongly inspired from those written by David Z Maze
for i2cset.


CHANGES
i2c-tools CHANGES
-----------------

4.3 (2021-07-22)
  decode-dimms: Attempt to decode LPDDR3 modules
  eeprom, eepromer: Removed the tools in favor of eeprog
  i2cdetect: Sort the bus list by number
  i2cdump: Add range support to I2C block mode
           Deprecate SMBus block mode
  i2cget: Add support for I2C block read
          Add support for SMBus block read
  i2ctransfer: Reverted check for returned length from driver

4.2 (2020-09-22)
  manpages: Add BUGS section to let people know how to contact us
  Makefile: Allow to preset all CFLAGS and LDFLAGS variables
  tools: Consistently use snprintf instead of sprintf
         Restrict addresses 0x03-0x07, too (defined by I2C standard)
  decode-dimms: Print SPD revision for DDR3 too
                Print primary bus width for DDR3 and DDR4
                List ee1004 as a candidate driver
                Display MAC for DDR3
                Add MAC abbreviation for DDR4
                Round DDR4 speed properly
                Detect and report truncated input files
                Print kernel driver used
                Print DDR memory speed in MT/s
                Add DDR5 memory types
                Decode manufacturing data for LPDDR3
                Fix the version string
                Point the user to the right drivers
                Update the list of vendors to Jedec JEP106BB
  decode-vaio: Add support for the at24 driver
               Scan more i2c buses
  i2cset: Fix short writes with mask
  i2ctransfer: Mention '-a' everywhere in the manpage
               Support messages using I2C_M_RECV_LEN
               Add check for returned length from driver
  i2c-stub-from-dump: Read dumps from hexdump -C
  library: Add a manual page to document the API

4.1 (2018-11-30)
  Makefile: Make STRIP, DESTDIR and PREFIX overridable
  tools: Fix potential buffer overflows in i2cbusses
         Fix build race
         Allow usage of reserved addresses with the '-a' flag
  decode-dimms: Add preliminary DDR4 support
                Decode size and timings of DDR4
                Decode misc parameters of DDR4
                Decode physical characteristics of DDR4
                Documentation update for DDR4
                Verify the CRC of DDR4 data block 1
                Update manufacturer IDs (JEP106AX)
  eeprog: Fix ambiguous parentheses
          Fix build race
  i2ctransfer: Rename option '-f' to '-a' for consistency
  i2c-dev.h: Delete
  library: Fix build race
           Allow disabling the dynamic flavor
           Mention the correct license in source files
  py-smbus: Fix i2c_smbus_* error propagation

4.0 (2017-10-30)
  tools: Fix build with recent compilers (gcc 4.6+)
         Add examples to the manual pages
  README: Clarify licenses
          Mention the current maintainer
  decode-dimms: Decode module configuration type of DDR2 SDRAM
                Decode bus width extension of DDR3 SDRAM
                Don't choke when no EEPROM is found
                Don't make columns larger than they need to be
                Make side-by-side output more robust
                Print module organization of DDR SDRAM
                Merge cells by default in side-by-side output
                Print extra timing values of DDR SDRAM
                Print DDR and DDR2 core timings for all supported CAS values
                Print DDR2 equivalent speed of tCK max
                Don't print undefined DDR2 SDRAM timings
		Print SDR, DDR, DDR2, DDR3 core timings for all standard speeds
                Update manufacturer IDs
                Make DDR3 manufacturer count parity error non-fatal
                Strip former manufacturer name in side-by-side output mode
                Remove duplicate "ns" in SDR timings
                Add section headers for SDR modules
                Fix decoding of SDR SPD revision
                Prevent hang on reserved DDR3 module type
                Decode more DDR3 module types
                Fix DDR3 tRAS decoding
                Fix DDR3 core timings rounding
                Round down PC3 numbers to comply with Jedec
                Don't print the DDR3 time bases
                Decode the FTB fields of DDR3 tCk, tAA, tRCD, tRP and tRC
                Fix speed and PC3 number of high-speed DDR3 modules
                Decode DDR3 reference card revision
                Print width of all known DDR3 module types
                Print physical characteristics for all DDR3 module types
                Don't print raw SSTE32882 register values
                Add support for Load Reduced DIMM (LRDIMM) DDR3 modules
                Fully decode the DDR3 SDRAM Device Type field
                Fix DDR3 extended temp range refresh rate decoding
                Encode "degrees" to HTML degree symbol
                Generate XHTML 1.1 compliant markup
                Add a manual page
                Correctly check for out-of-bounds vendor ID
                Update manufacturer IDs (JEP106AQ)
  decode-vaio: Add a manual page
  eeprog: Add a manual page
          Moved to a separate subdirectory
          Increase delay after writes
  eeprom: Add a manual page
          Marked as deprecated
  eepromer: Add a manual page
            Marked as deprecated
  i2cdetect: Do a best effort detection if functionality is missing
             Clarify the SMBus commands used for probing by default
  i2ctransfer: New tool to send user-defined I2C messages in one transfer
  i2c-dev.h: Minimize differences with kernel flavor
             Move SMBus helper functions to include/i2c/smbus.h
  i2c-stub-from-dump: Be more tolerant on input dump format
  library: New libi2c library
           Properly propagate real error codes on read errors
           Use I2C_SMBUS_BLOCK_MAX instead of hard-coding 32
  lib/smbus.c: Add missing include which was causing a build error
  py-smbus: Fix module level docs
            Add support for python 3

3.1.0 (2011-12-04)

tools/i2cget.c
/*
    i2cget.c - A user-space program to read an I2C register.
    Copyright (C) 2005-2021  Jean Delvare <jdelvare@suse.de>

    Based on i2cset.c:
    Copyright (C) 2001-2003  Frodo Looijaard <frodol@dds.nl>, and
                             Mark D. Studebaker <mdsxyz123@yahoo.com>
    Copyright (C) 2004-2005  Jean Delvare

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
    MA 02110-1301 USA.
*/

#include <sys/ioctl.h>
#include <errno.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <linux/i2c.h>
#include <linux/i2c-dev.h>
#include <i2c/smbus.h>
#include "i2cbusses.h"
#include "util.h"
#include "../version.h"

static void help(void) __attribute__ ((noreturn));

static void help(void)
{
	fprintf(stderr,
		"Usage: i2cget [-f] [-y] [-a] I2CBUS CHIP-ADDRESS [DATA-ADDRESS [MODE [LENGTH]]]\n"
		"  I2CBUS is an integer or an I2C bus name\n"
		"  ADDRESS is an integer (0x08 - 0x77, or 0x00 - 0x7f if -a is given)\n"
		"  MODE is one of:\n"
		"    b (read byte data, default)\n"
		"    w (read word data)\n"
		"    c (write byte/read byte)\n"
		"    s (read SMBus block data)\n"
		"    i (read I2C block data)\n"
		"    Append p for SMBus PEC\n"
		"  LENGTH is the I2C block data length (between 1 and %d, default %d)\n",
		I2C_SMBUS_BLOCK_MAX, I2C_SMBUS_BLOCK_MAX);
	exit(1);
}

static int check_funcs(int file, int size, int daddress, int pec)
{
	unsigned long funcs;

	/* check adapter functionality */
	if (ioctl(file, I2C_FUNCS, &funcs) < 0) {
		fprintf(stderr, "Error: Could not get the adapter "
			"functionality matrix: %s\n", strerror(errno));
		return -1;
	}

	switch (size) {
	case I2C_SMBUS_BYTE:
		if (!(funcs & I2C_FUNC_SMBUS_READ_BYTE)) {
			fprintf(stderr, MISSING_FUNC_FMT, "SMBus receive byte");
			return -1;
		}
		if (daddress >= 0
		 && !(funcs & I2C_FUNC_SMBUS_WRITE_BYTE)) {
			fprintf(stderr, MISSING_FUNC_FMT, "SMBus send byte");
			return -1;
		}
		break;

	case I2C_SMBUS_BYTE_DATA:
		if (!(funcs & I2C_FUNC_SMBUS_READ_BYTE_DATA)) {
			fprintf(stderr, MISSING_FUNC_FMT, "SMBus read byte");
			return -1;
		}
		break;

	case I2C_SMBUS_WORD_DATA:
		if (!(funcs & I2C_FUNC_SMBUS_READ_WORD_DATA)) {
			fprintf(stderr, MISSING_FUNC_FMT, "SMBus read word");
			return -1;
		}
		break;

	case I2C_SMBUS_BLOCK_DATA:
		if (!(funcs & I2C_FUNC_SMBUS_READ_BLOCK_DATA)) {
			fprintf(stderr, MISSING_FUNC_FMT, "SMBus block read");
			return -1;
		}
		break;

	case I2C_SMBUS_I2C_BLOCK_DATA:
		if (!(funcs & I2C_FUNC_SMBUS_READ_I2C_BLOCK)) {
			fprintf(stderr, MISSING_FUNC_FMT, "I2C block read");
			return -1;
		}
		break;
	}

	if (pec
	 && !(funcs & (I2C_FUNC_SMBUS_PEC | I2C_FUNC_I2C))) {
		fprintf(stderr, "Warning: Adapter does "
			"not seem to support PEC\n");
	}

	return 0;
}

static int confirm(const char *filename, int address, int size, int daddress,
		   int length, int pec)
{
	int dont = 0;

	fprintf(stderr, "WARNING! This program can confuse your I2C "
		"bus, cause data loss and worse!\n");

	/* Don't let the user break his/her EEPROMs */
	if (address >= 0x50 && address <= 0x57 && pec) {
		fprintf(stderr, "STOP! EEPROMs are I2C devices, not "
			"SMBus devices. Using PEC\non I2C devices may "
			"result in unexpected results, such as\n"
			"trashing the contents of EEPROMs. We can't "
			"let you do that, sorry.\n");
		return 0;
	}

	if (size == I2C_SMBUS_BYTE && daddress >= 0 && pec) {
		fprintf(stderr, "WARNING! All I2C chips and some SMBus chips "
			"will interpret a write\nbyte command with PEC as a"
			"write byte data command, effectively writing a\n"
			"value into a register!\n");
		dont++;
	}

	fprintf(stderr, "I will read from device file %s, chip "
		"address 0x%02x, ", filename, address);
	if (daddress < 0)
		fprintf(stderr, "current data\naddress");
	else
		fprintf(stderr, "data address\n0x%02x", daddress);
	if (size == I2C_SMBUS_I2C_BLOCK_DATA)
		fprintf(stderr, ", %d %s using read I2C block data.\n",
			length, length > 1 ? "bytes" : "byte");
	else
		fprintf(stderr, ", using %s.\n",
			size == I2C_SMBUS_BYTE ? (daddress < 0 ?
			"read byte" : "write byte/read byte") :
			size == I2C_SMBUS_BYTE_DATA ? "read byte data" :
			size == I2C_SMBUS_BLOCK_DATA ? "read SMBus block data" :
			"read word data");
	if (pec)
		fprintf(stderr, "PEC checking enabled.\n");

	fprintf(stderr, "Continue? [%s] ", dont ? "y/N" : "Y/n");
	fflush(stderr);
	if (!user_ack(!dont)) {
		fprintf(stderr, "Aborting on user request.\n");
		return 0;
	}

	return 1;
}

int main(int argc, char *argv[])
{
	char *end;
	int res, i2cbus, address, size, file;
	int daddress;
	char filename[20];
	int pec = 0;
	int flags = 0;
	int force = 0, yes = 0, version = 0, all_addrs = 0;
	int length;
	unsigned char block_data[I2C_SMBUS_BLOCK_MAX];

	/* handle (optional) flags first */
	while (1+flags < argc && argv[1+flags][0] == '-') {
		switch (argv[1+flags][1]) {
		case 'V': version = 1; break;
		case 'f': force = 1; break;
		case 'y': yes = 1; break;
		case 'a': all_addrs = 1; break;
		default:
			fprintf(stderr, "Error: Unsupported option "
				"\"%s\"!\n", argv[1+flags]);
			help();
			exit(1);
		}
		flags++;
	}

	if (version) {
		fprintf(stderr, "i2cget version %s\n", VERSION);
		exit(0);
	}

	if (argc < flags + 3)
		help();

	i2cbus = lookup_i2c_bus(argv[flags+1]);
	if (i2cbus < 0)
		help();

	address = parse_i2c_address(argv[flags+2], all_addrs);
	if (address < 0)
		help();

	if (argc > flags + 3) {
		size = I2C_SMBUS_BYTE_DATA;
		daddress = strtol(argv[flags+3], &end, 0);
		if (*end || daddress < 0 || daddress > 0xff) {
			fprintf(stderr, "Error: Data address invalid!\n");
			help();
		}
	} else {
		size = I2C_SMBUS_BYTE;
		daddress = -1;
	}

	if (argc > flags + 4) {
		switch (argv[flags+4][0]) {
		case 'b': size = I2C_SMBUS_BYTE_DATA; break;
		case 'w': size = I2C_SMBUS_WORD_DATA; break;
		case 'c': size = I2C_SMBUS_BYTE; break;
		case 's': size = I2C_SMBUS_BLOCK_DATA; break;
		case 'i': size = I2C_SMBUS_I2C_BLOCK_DATA; break;
		default:
			fprintf(stderr, "Error: Invalid mode!\n");
			help();
		}
		pec = argv[flags+4][1] == 'p';
		if (size == I2C_SMBUS_I2C_BLOCK_DATA && pec) {
			fprintf(stderr, "Error: PEC not supported for I2C block data!\n");
			help();
		}
	}

	if (argc > flags + 5) {
		if (size != I2C_SMBUS_I2C_BLOCK_DATA) {
			fprintf(stderr, "Error: Length only valid for I2C block data!\n");
			help();
		}
		length = strtol(argv[flags+5], &end, 0);
		if (*end || length < 1 || length > I2C_SMBUS_BLOCK_MAX) {
			fprintf(stderr, "Error: Length invalid!\n");
			help();
		}
	} else {
		length = I2C_SMBUS_BLOCK_MAX;
	}

	file = open_i2c_dev(i2cbus, filename, sizeof(filename), 0);
	if (file < 0
	 || check_funcs(file, size, daddress, pec)
	 || set_slave_addr(file, address, force))
		exit(1);

	if (!yes && !confirm(filename, address, size, daddress, length, pec))
		exit(0);

	if (pec && ioctl(file, I2C_PEC, 1) < 0) {
		fprintf(stderr, "Error: Could not set PEC: %s\n",
			strerror(errno));
		close(file);
		exit(1);
	}

	switch (size) {
	case I2C_SMBUS_BYTE:
		if (daddress >= 0) {
			res = i2c_smbus_write_byte(file, daddress);
			if (res < 0)
				fprintf(stderr, "Warning - write failed\n");
		}
		res = i2c_smbus_read_byte(file);
		break;
	case I2C_SMBUS_WORD_DATA:
		res = i2c_smbus_read_word_data(file, daddress);
		break;
	case I2C_SMBUS_BLOCK_DATA:
		res = i2c_smbus_read_block_data(file, daddress, block_data);
		break;
	case I2C_SMBUS_I2C_BLOCK_DATA:
		res = i2c_smbus_read_i2c_block_data(file, daddress, length, block_data);
		break;
	default: /* I2C_SMBUS_BYTE_DATA */
		res = i2c_smbus_read_byte_data(file, daddress);
	}
	close(file);

	if (res < 0) {
		fprintf(stderr, "Error: Read failed\n");
		exit(2);
	}

	if (size == I2C_SMBUS_BLOCK_DATA ||
	    size == I2C_SMBUS_I2C_BLOCK_DATA) {
		int i;

		for (i = 0; i < res - 1; ++i)
			printf("0x%02hhx ", block_data[i]);
		printf("0x%02hhx\n", block_data[res - 1]);
	} else {
		printf("0x%0*x\n", size == I2C_SMBUS_WORD_DATA ? 4 : 2, res);
	}

	exit(0);
}
